module/zfs/vdev.c

```diff
@@ -3819,17 +3819,6 @@ vdev_stat_update(zio_t *zio, uint64_t psize)
 	if (zio->io_vd == NULL && (zio->io_flags & ZIO_FLAG_DONT_PROPAGATE))
 		return;
 
-	mutex_enter(&vd->vdev_stat_lock);
-	if (type == ZIO_TYPE_READ && !vdev_is_dead(vd)) {
-		if (zio->io_error == ECKSUM)
-			vs->vs_checksum_errors++;
-		else
-			vs->vs_read_errors++;
-	}
-	if (type == ZIO_TYPE_WRITE && !vdev_is_dead(vd))
-		vs->vs_write_errors++;
-	mutex_exit(&vd->vdev_stat_lock);
```

Comment:
I'm concerned that moving this logic is going to result in some soft errors being reported as hard errors.  For example, speculative IOs and which can legitimately race and retries which should be suppressed.

Reply:
The new logic relies on when IO/checksum events are posted to increment the counters.  If we trust that the code in `master` is sending the correct events, then the counters should be correct as well.  I do see that there's some zio/vdev sanity checks in `zfs_ereport_start()` so maybe the bogus events are getting caught there.